src/main/scala/util/RunUtils.scala

```diff
@@ -120,105 +144,167 @@ object RunUtils {
-    //print("Memory Model Map: \n")
-    //print(mmm)
-    val interprocCfg = InterproceduralProgramCfg.generateFromProgram(IRProgram)
-    Output.output(OtherOutput(OutputKindE.cfg), interprocCfg.toDot(x => x.toString, Output.dotIder), "inter_cfg")
+    mmm.convertMemoryRegions(result2, mergedSubroutines)
+
+    Logger.info("[!] Running VSA")
+    val solver3 =
+      ValueSetAnalysis.WorklistSolver(cfg, globalAddresses, externalAddresses, globalOffsets, subroutines, mmm, result)
+    val result3 = solver3.analyze(false)
+    Output.output(
+      OtherOutput(OutputKindE.cfg),
+      cfg.toDot(Output.labeler(result3, solver3.stateAfterNode), Output.dotIder),
+      "vsa"
+    )
+
+    Logger.info("[!] Resolving CFG")
+    val (newIR, modified) = resolveCFG(cfg, result3.asInstanceOf[Map[CfgNode, Map[Expr, Set[Value]]]], IRProgram)
+    if (modified) {
+      Logger.info(s"[!] Analysing again (iter $iterations)")
+      return analyse(newIR, externalFunctions, globals, globalOffsets)
+    }
+    val newCFG = ProgramCfgFactory().fromIR(newIR, inlineLimit = 0)
+    Output.output(OtherOutput(OutputKindE.cfg), newCFG.toDot(x => x.toString, Output.dotIder), "resolvedCFG")
 
-    val solver3 = ValueSetAnalysis.WorklistSolver(interprocCfg, globalAddresses, externalAddresses, globalOffsets, subroutines, mmm)
-    val result3 = solver3.analyze()
-    Output.output(OtherOutput(OutputKindE.cfg), interprocCfg.toDot(Output.labeler(result3, solver3.stateAfterNode), Output.dotIder), "vsa")
+    Logger.info(s"[!] Finished indirect call resolution after $iterations iterations")
 
-    val newCFG = InterproceduralProgramCfg.generateFromProgram(resolveCFG(interprocCfg, result3.asInstanceOf[Map[CfgNode, Map[Expr, Set[Value]]]], IRProgram))
-    Output.output(OtherOutput(OutputKindE.cfg), newCFG.toDot(x => x.toString, Output.dotIder), "resolvedCFG")
+    newIR
   }
 
-  def resolveCFG(interproceduralProgramCfg: InterproceduralProgramCfg, valueSets: Map[CfgNode, Map[Expr, Set[Value]]], IRProgram: Program): Program = {
-    // print the count of the value sets of the exit nodes
-//    for (comdNode <- interproceduralProgramCfg.nodes.filter(_.isInstanceOf[CfgCommandNode])) {
-////      if (comdNode.asInstanceOf[CfgCommandNode].data.isInstanceOf[IndirectCall]) {
-////        //println(s"Node: ${comdNode}")
-////        println(s"${valueSets(comdNode).size}")
-////      }
-//      println(s"${valueSets(comdNode).size}")
-//    }
-    interproceduralProgramCfg.entries.foreach(n => process(n))
+  def resolveCFG(
+      cfg: ProgramCfg,
+      valueSets: Map[CfgNode, Map[Expr, Set[Value]]],
+      IRProgram: Program
+  ): (Program, Boolean) = {
+    var modified: Boolean = false
+    val worklist = ListBuffer[CfgNode]()
+    // find the main function
+    cfg.nodes.foreach {
+      case main: CfgFunctionEntryNode if main.data.name == "main" =>
+        main.succ(true).toSet.union(main.succ(false).toSet).foreach(node => worklist.addOne(node))
+      case _ =>
+    }
```

Comment:
This should use Program.mainProcedure to find the starting procedure now, instead of just finding the name "main"